### ISSUE TITLE
Evolved data model layer

### ISSUE THREAD

**Issue by Olaktal** [metabase/metabase#14061]
For now, we can query databases without knowing SQL thanks to Custom questions (which is possible thanks to metabase analyzing tables metadata) but our business users have still some pains because:
- they have to know the database schemas (for joining tables)
- finding a field for business users might still be a bit messy because fields are stored by tables which sometimes is not logical for business users

So what we are looking for is an abstract data model layer which could allow us to:
- abstract joins between tables
- reorganize data available following business issues

To do so we could define:

1. "views" which correspond to tables where we could choose the fields to display (or not), apply SQL transformation (Ex: Concat(field1,'abc')) and chosing a named folder/group where the field will be displayed/available for end users
2. "models" which describe how tables are linked together (type of join (left, right, ..), relationship (Ex: many_to_one, ..), fields, ..)

There could be multiple data model layers.

We could then have a drag and drop system based on these data model layers, SQL would be automatically generated from object choosed from the end user interface.
This could be another question type in metabase.

What I am describing is simply how evolved non-free analytics software such as Tableau or Looker (see [lookml](https://docs.looker.com/data-modeling/learning-lookml/what-is-lookml)) are basically working. For me, this is the main metabase missing feature that force us to pay (a lot) for these kind of solutions, leaving metabase.

**Comment by flamber** [comment]
Hi @Olaktal 
Sounds like this might be a duplicate of #2214 and #7335 and #14054 - and very related to #1708 and #3475.
Or am I misunderstanding what you're looking for?

**Comment by Olaktal** [comment]
Hi @flamber, 
It doesn't seem like these tickets would totaly abstract the database schema as I do, preventing end users from knowing tables joins or whatever technical info related to the database.
Here is an interesting feature list extracted from these tickets: 
- #2214: no more joining for end users and pre-set columns transformations
- #12800: pivot function (transform rows to column) -> nice to have but in a second time
- #7335: join 2 tables on multiple fields
- #14054: join a table to different tables using the same field depending on a second field value (aka "polymorphic foreign key relationships")

I agree with you that the solution I propose would answer all of these features. My proposal is just more detailed and would be a solid base to "easily" implement these features. Indeed #2214, #7335 and #14054 could be implemented in the model layer

We would need a declarative configuration language (kind of HCL for terraform) to define the data model layer.
In the admin panel, we would first have to declare each views that point to tables and then declare the model that describe tables joins.

Exemple with 3 tables: table_a, table_b and table_c

``` #view tableA

view: tableA {
  	db_table_name: schema.table_a

	field: column1_var {
	  	sql: column1
	  	hide: True
	  }

	field: column2_var {
	  	sql: column2
	  	hide: True
  	}

	field: column3_var {
	  	label: "column 3 label"
	  	folder_label: "Entity info"
	  	sql: column3
	  	hide: False
  	}

	field: column4_var {
	  	label: "column 4 label"
	  	folder_label: "Entity info"
	  	sql: column4
	  	hide: False
  	}

	field: column5_var {
	  	label: "column 5 label"
	  	folder_label: "Other folder"
	  	sql: CONCAT(column5, "ABC")
	  	hide: False
  	}

  	...
```

``` #view tableB

view: tableB {
  	db_table_name: schema.table_b

	field: column2_var {
	  	sql: column2
	  	hide: True
  	}

	field: column3_var {
	  	label: "column 3 label"
	  	folder_label: "Other folder 2 "
	  	sql: column3
	  	hide: False
  	}

  	...
```


``` #view tableC

view: tableC {
  	db_table_name: schema.table_c

	field: column2_var {
	  	sql: column2
	  	hide: True
  	}

	field: column3_var {
	  	label: "column 3 label"
	  	folder_label: "Other folder 3 "
	  	sql: column3
	  	hide: False
  	}

  	...
```

And then the model where we join views

``` #modelA definition

model : tableA {
  label: "Model name"

  join : tableB {
    type: left
    relationship: many_to_one
    sql: ${tableA.column1_var} = ${tableB.column2_var}
  }

  join : tableC {
    type: left
    relationship: many_to_one
    sql: ${tableA.column2_var} = ${tableC.column2_var} AND ${tableA.column3_var} = 'entityType'
  }
```

For end/business users, each "object" declared in the different views would appear in the "create a new question with drag and drop" interface.
The declared model/views would automatically translate end user actions to SQL query.

**Comment by andreychirikba** [comment]
We do such things with regular SQL views, denormalizing a lot of tables into a single view, works like a charm.

**Comment by Olaktal** [comment]
@andreychirikba good idea ! It can be useful when you know what you need and what people would looking for but this might more fit the usage for stabilized company than small fast growing company.
Also joins are always made between tables inside a view, my solution would join only tables where a column is called/needed

**Comment by thebiglabasky** [comment]
This seems to be doable using Models now, so I'm tempted to close this

**Comment by thebiglabasky** [comment]
Feel free to reopen if I'm missing something
